pkg/3scale/amp/operator/persistent_volume_claim_reconciler_test.go

```diff
@@ -0,0 +1,224 @@
+	createOnlyReconciler := NewCreateOnlyPVCReconciler()
+
+	reconciler := NewPVCBaseReconciler(baseAPIManagerLogicReconciler, createOnlyReconciler)
+
+	desired := &v1.PersistentVolumeClaim{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "v1",
+			Kind:       "PersistentVolumeClaim",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "myPVC",
+			Namespace: namespace,
+		},
+	}
+
+	err := reconciler.Reconcile(desired)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	namespacedName := types.NamespacedName{
+		Name:      "myPVC",
+		Namespace: namespace,
+	}
+	existing := &v1.PersistentVolumeClaim{}
+	err = cl.Get(context.TODO(), namespacedName, existing)
+	// object must exist, that is all required to be tested
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestPVCBaseReconcilerUpdateOwnerRef(t *testing.T) {
+	var (
+		name      = "example-apimanager"
+		namespace = "operator-unittest"
+		log       = logf.Log.WithName("operator_test")
+	)
+	apimanager := &appsv1alpha1.APIManager{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
+		Spec: appsv1alpha1.APIManagerSpec{},
+	}
+	existing := &v1.PersistentVolumeClaim{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "v1",
+			Kind:       "PersistentVolumeClaim",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "myPVC",
+			Namespace: namespace,
+		},
+	}
+	s := scheme.Scheme
+	s.AddKnownTypes(appsv1alpha1.SchemeGroupVersion, apimanager)
+
+	// Objects to track in the fake client.
+	objs := []runtime.Object{existing}
+
+	// Create a fake client to mock API calls.
+	cl := fake.NewFakeClient(objs...)
+	clientAPIReader := fake.NewFakeClient(objs...)
+
+	baseReconciler := NewBaseReconciler(cl, clientAPIReader, s, log)
+	baseLogicReconciler := NewBaseLogicReconciler(baseReconciler)
+	baseAPIManagerLogicReconciler := NewBaseAPIManagerLogicReconciler(baseLogicReconciler, apimanager)
+	reconciler := NewPVCBaseReconciler(baseAPIManagerLogicReconciler, NewCreateOnlyPVCReconciler())
+
+	desired := existing.DeepCopy()
+
+	err := reconciler.Reconcile(desired)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	namespacedName := types.NamespacedName{
+		Name:      "myPVC",
+		Namespace: namespace,
+	}
+
+	reconciled := &v1.PersistentVolumeClaim{}
+	err = cl.Get(context.TODO(), namespacedName, reconciled)
+	// object must exist, that is all required to be tested
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if len(reconciled.GetOwnerReferences()) != 1 {
+		t.Fatal("reconciled does not have owner reference")
+	}
+
+	if reconciled.GetOwnerReferences()[0].Name != name {
+		t.Fatalf("reconciled owner reference is not apimanager, expected: %s, got: %s", name, reconciled.GetOwnerReferences()[0].Name)
+	}
+}
+
+type myCustomPVCReconciler struct {
+}
+
+func (r *myCustomPVCReconciler) IsUpdateNeeded(desired, existing *v1.PersistentVolumeClaim) bool {
+	tmp := "myClass"
+	existing.Spec.StorageClassName = &tmp
```

Comment:
Although it's true that we are reconciling the value of StorageClass, PVCs do not support changing their StorageClassName value once the PVC is created and bound.

Trying to edit the field results i nthe following message:
```
# * spec: Forbidden: is immutable after creation except resources.requests for bound claims
```

Reply:
For PVC we are only using CreateOnly reconcilers, so we are no reconciling StorageClass (if I am not wrong).

In this test, the purpose was to check that PVC reconciler updates if reconciler type is not CreateOnly.